### PR TITLE
upgrade to declarative setuptools

### PR DESCRIPTION
**Pull request by graingert**
## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/9999
* [ ] I ran `tox -e black-reformat` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [ ] I ran additional checks listed at [Getting Your Patch Accepted](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#GettingYourPatchAccepted)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

### CODE REVIEW

File: src/twisted/python/_setup.py
Comment (graingert) [inline src/twisted/python/_setup.py:182]: currently _EXTRAS_REQUIRE is in python to save on duplication eg with `twisted[all_non_platform]`

File: setup.cfg
Comment (mthuurne) [inline setup.cfg:21]: 3.9 was released this week and CI passes on 3.9-rc2, so I think it's safe to add 3.9 to the classifiers as well.

File: setup.cfg
Comment (mthuurne) [inline setup.cfg]: Since `typing.TYPE_CHECKING` was added in 3.5.2, I think the minimum version needs a tiny bump.

File: src/twisted/python/_setup.py
Comment (mthuurne) [inline src/twisted/python/_setup.py]: Since `readme` contains a path, why not make `Path` its type?

File: src/twisted/python/test/test_setup.py
Comment (mthuurne) [inline src/twisted/python/test/test_setup.py]: Since `getSetupArgs()` just passes on the result of `_longDescriptionFromReadme()` unchanged, you could instead call the latter directly. That avoids this test case creating an inflated amount of coverage.

File: src/twisted/python/test/test_setup.py
Comment (mthuurne) [inline src/twisted/python/test/test_setup.py]:
```suggestion
        path.touch(exist_ok=False)
```

**Review by mthuurne — COMMENTED** [review]
I added some comments suggesting potential cleanups. Overall it looks good.

I only reviewed the code by reading; I didn't test whether this produces a working wheel with all the correct metadata.

One concern I have though is that `test_mail.py` and `test_options.py` are removed, but the code that they're testing is not. This is probably related to the removal of the `notPortedModules` list, but it seems like something that should be resolved in a more thorough matter, perhaps in a separate PR.

File: src/twisted/python/test/test_setup.py
Comment (graingert) [reply src/twisted/python/test/test_setup.py]: this is intentional to avoid coupling getSetupArgs to its specific implementation

File: src/twisted/python/test/test_setup.py
Comment (graingert) [inline src/twisted/python/test/test_setup.py]:
```suggestion
        path.touch(exist_ok=False)
```

File: setup.cfg
Comment (graingert) [reply setup.cfg]: doesn't this stuff get backported without changing the version in some platforms?

I think it will be changed to 3.6 soon though

File: src/twisted/python/test/test_setup.py
Comment (graingert) [reply src/twisted/python/test/test_setup.py]: now that _longDescriptionFromReadme is one expression - I refactored it into getSetupArgs

File: setup.cfg
Comment (graingert) [reply setup.cfg:21]:
```suggestion
    Programming Language :: Python :: 3.8
    Programming Language :: Python :: 3.9
```

**Comment by graingert** [comment]
> One concern I have though is that test_mail.py and test_options.py are removed, but the code that they're testing is not. This is probably related to the removal of the notPortedModules list, but it seems like something that should be resolved in a more thorough matter, perhaps in a separate PR.

I think reverting the tests and fixing them is on the table. Or deleting these `notPortedModules`. I don't think they do any harm in the dist though

File: setup.cfg
Comment (mthuurne) [reply setup.cfg]: - With `>=3.5` on a system that runs actual Python 3.5.0, `pip` will install the new Twisted release and then Twisted will break when importing `typing.TYPE_CHECKING`.
- With `>=3.5.2` on a system that runs Python that identifies as 3.5.0 but is actually patched with bits from later releases, `pip` won't install the new Twisted release.

Since the latter is the least disruptive for the user, I think it's preferable.

There are plans to drop 3.5, see #1386, but that isn't scheduled to be merged until after the next release.

Comment (graingert) [reply setup.cfg]: I've pushed that change now

**Comment by wsanchez** [comment]
> > One concern I have though is that test_mail.py and test_options.py are removed, but the code that they're testing is not. This is probably related to the removal of the notPortedModules list, but it seems like something that should be resolved in a more thorough matter, perhaps in a separate PR.
> 
> I think reverting the tests and fixing them is on the table. Or deleting these `notPortedModules`. I don't think they do any harm in the dist though

I agree with @mthuurne that removing them is problematic…

**Comment by wsanchez** [comment]
WHOAH oops didn't mean to close

**Comment by wsanchez** [comment]
I think this is a lot more readable and I like moving the static config back up to the top of the project.  I've never really used `setup.cfg` so I'm not aware of what pitfalls may exist, if any, or what the best way to verify the built result is OK, though I guess that's what pre-releases are for…?

**Comment by graingert** [comment]
> or what the best way to verify the built result is OK, though I guess that's what pre-releases are for…?

a pypi pre-release or a test.pypi.org upload would work. Tox does run this so it mostly works and the wheels seem fine.

you can run `python setup.py sdist bdist_wheel` and inspect the results

**Comment by wsanchez** [comment]
> a pypi pre-release or a test.pypi.org upload would work. Tox does run this so it mostly works and the wheels seem fine.

I like that.

@rodrigc @glyph How do you feel about giving one of us upload permissions to Twisted on `test.pypi.org`?

**Comment by graingert** [comment]
@wsanchez you could rename the `setup.cfg[metadata].name` key

**Comment by rodrigc** [comment]
> @rodrigc @glyph How do you feel about giving one of us upload permissions to Twisted on `test.pypi.org`?

I can't help you with that.  I also am blocked on that: https://github.com/twisted/twisted/pull/1423#issuecomment-703744080

**Comment by graingert** [comment]
@rodrigc you appear to be soul maintainer of https://test.pypi.org/project/Twisted/ you could add @wsanchez as a maintainer

**Comment by rodrigc** [comment]
Interesting.  I don't know when I was granted that access to test.pypi.org.  For now I am going to focus on moving forward with the release, before giving additional people access to PyPI resources.  If Glyph or Hawkowl want to grant access, I will defer to their wishes.

**Comment by glyph** [comment]
@rodrigc You weren't "granted access" - you own the project in its entirety on test.pypi.org; the data there does not derive from PyPI at all.

https://test.pypi.org/project/Twisted/

I didn't even have an account on test.pypi.org — I just made one, so at least I got "glyph" :).  Please go ahead and give everybody who needs it access to test PyPI (including me :)).

**Comment by rodrigc** [comment]
@glyph OK, I granted you and @wsanchez access to https://test.pypi.org/ .  I couldn't find an account for hawkowl on test.pypi.org to grant access.

**Comment by graingert** [comment]
> I agree with @mthuurne that removing them is problematic…

I've restored the tests that were "commented out" in src/twisted/python/_setup.py, ported just enough code to get them to run the test setUps and skipped every test that fails

**Review by wsanchez — APPROVED** [review]
cool, thanks

**Pull request merged**
